# **rAcecaR: Antimicrobial Resistance (AR) Bioinformatics Pipeline**

[![Hits](https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fgithub.com%2FODHL%2FODHL_AR%2F&count_bg=%233D4FC8&title_bg=%23555555&icon=&icon_color=%23E7E7E7&title=PageHits&edge_flat=false)](https://hits.seeyoufarm.com)

ðŸš¨ **Warning** ðŸš¨  
The results generated by this pipeline are **not CLIA certified** and should not be considered diagnostic.

---

## **Introduction**  
The **ODHL_rAcecaR pipeline** is designed for the detection and characterization of antimicrobial resistance (AMR) genes and other relevant genomic features from high-throughput sequencing data. This pipeline processes bacterial whole genome sequencing (WGS) and metagenomic sequencing data to identify antimicrobial resistance determinants, sequence types, and genomic features relevant to public health surveillance and epidemiological investigations.

This workflow is developed and maintained by the [`Ohio Department of Health Laboratory (ODHL)`](https://odh.ohio.gov/about-us/offices-bureaus-and-departments/bphl/welcome-to?contentIDR=9ce211e2-7949-447b-bf0e-2abfd5da7892&useDefaultText=1&useDefaultDesc=0) and is built using [`Nextflow`](https://www.nextflow.io/), a robust workflow management system that ensures reproducibility and scalability across HPC clusters, cloud environments, and local machines. The pipeline was built off the work of the [`CDC Phoenix Pipeline`](https://github.com/CDCgov/phoenix). The workflow leverages [`Docker`](https://www.docker.com/) and [`Singularity containers`](https://www.singularity.com/) to provide a standardized execution environment.
[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A521.10.3-23aa62.svg?labelColor=000000)](https://www.nextflow.io/)

---

## **Pipeline Information**
View our [`docs`](https://odhl.github.io/ODHL_AR/) website for more information on the pipeline, how to execute, input requirements, output files, and contributions.