<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content=ODHL/rAcecaR><meta name=author content="Ohio Department of Health Laboratory"><link rel=prev href=../..><link href=../preparing-files/ rel=next><link rel=icon href=../../assets/images/favicon.png><meta name=generator content="mkdocs-1.5.3, mkdocs-material-9.1.6"><title>1. Getting Started - ODHL/rAcecaR</title><link rel=stylesheet href=../../assets/stylesheets/main.ded33207.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.a0c5b2b5.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#getting-started class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=ODHL/rAcecaR class="md-header__button md-logo" aria-label=ODHL/rAcecaR data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> ODHL/rAcecaR </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> 1. Getting Started </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/ODHL/rAcecaR title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> ODHL/rAcecaR </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> Background </a> </li> <li class=md-tabs__item> <a href=./ class="md-tabs__link md-tabs__link--active"> Usage </a> </li> <li class=md-tabs__item> <a href=../troubleshooting/ class=md-tabs__link> FAQ </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=ODHL/rAcecaR class="md-nav__button md-logo" aria-label=ODHL/rAcecaR data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> ODHL/rAcecaR </label> <div class=md-nav__source> <a href=https://github.com/ODHL/rAcecaR title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> ODHL/rAcecaR </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> Background </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> Usage <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Usage </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> 1. Getting Started <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> 1. Getting Started </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#introduction class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=#pipeline-summary class=md-nav__link> Pipeline Summary </a> </li> <li class=md-nav__item> <a href=#entry-points class=md-nav__link> Entry Points </a> </li> <li class=md-nav__item> <a href=#processes class=md-nav__link> Processes </a> <nav class=md-nav aria-label=Processes> <ul class=md-nav__list> <li class=md-nav__item> <a href=#quality-control-preprocessing class=md-nav__link> Quality Control &amp; Preprocessing </a> </li> <li class=md-nav__item> <a href=#genome-assembly-assessment class=md-nav__link> Genome Assembly &amp; Assessment </a> </li> <li class=md-nav__item> <a href=#genotyping-comparative-genomics class=md-nav__link> Genotyping &amp; Comparative Genomics </a> </li> <li class=md-nav__item> <a href=#antimicrobial-resistance-virulence-detection class=md-nav__link> Antimicrobial Resistance &amp; Virulence Detection </a> </li> <li class=md-nav__item> <a href=#genome-annotation class=md-nav__link> Genome Annotation </a> </li> <li class=md-nav__item> <a href=#final-reports class=md-nav__link> Final Reports </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#dependencies class=md-nav__link> Dependencies </a> <nav class=md-nav aria-label=Dependencies> <ul class=md-nav__list> <li class=md-nav__item> <a href=#software-tools class=md-nav__link> Software &amp; Tools </a> </li> <li class=md-nav__item> <a href=#core-software-versions class=md-nav__link> Core Software Versions </a> </li> <li class=md-nav__item> <a href=#reference-databases class=md-nav__link> Reference Databases </a> <nav class=md-nav aria-label="Reference Databases"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#kraken2-db class=md-nav__link> Kraken2 DB </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#pipeline-execution class=md-nav__link> Pipeline Execution </a> <nav class=md-nav aria-label="Pipeline Execution"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-install-nextflow class=md-nav__link> 1. Install Nextflow </a> </li> <li class=md-nav__item> <a href=#2-clone-the-repository class=md-nav__link> 2. Clone the Repository </a> </li> <li class=md-nav__item> <a href=#3-configure-the-pipeline class=md-nav__link> 3. Configure the Pipeline </a> </li> <li class=md-nav__item> <a href=#4-optional-install-basespace class=md-nav__link> *** 4. (Optional) Install Basespace** </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#reproducibility class=md-nav__link> Reproducibility </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../preparing-files/ class=md-nav__link> 2. Preparing Files </a> </li> <li class=md-nav__item> <a href=../run/ class=md-nav__link> 3. Running the Pipeline </a> </li> <li class=md-nav__item> <a href=../output/ class=md-nav__link> 4. Expected Output </a> </li> <li class=md-nav__item> <a href=../test-info/ class=md-nav__link> 5. Running Test Data </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> FAQ <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../troubleshooting/ class=md-nav__link> Troubleshooting </a> </li> <li class=md-nav__item> <a href=../contributions/ class=md-nav__link> Contributions </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=getting-started>Getting Started<a class=headerlink href=#getting-started title="Permanent link">&para;</a></h1> <h2 id=introduction>Introduction<a class=headerlink href=#introduction title="Permanent link">&para;</a></h2> <p><strong>ODHL_AR</strong> is a bioinformatics best-practice pipeline for detecting antimicrobial resistance (AMR) genes and assessing bacterial genomic characteristics from whole-genome sequencing (WGS) data.</p> <p>The pipeline is built using <a href=https://www.nextflow.io>Nextflow</a>, a workflow tool that efficiently runs tasks across multiple compute infrastructures in a portable and scalable manner. It uses Docker/Singularity containers, making installation trivial and ensuring reproducible results.</p> <hr> <h2 id=pipeline-summary>Pipeline Summary<a class=headerlink href=#pipeline-summary title="Permanent link">&para;</a></h2> <ol> <li><strong>Read QC</strong>: <a href=https://www.bioinformatics.babraham.ac.uk/projects/fastqc/ ><code>FastQC</code></a> – Provides quality metrics for raw sequencing reads.</li> <li><strong>Trimming reads</strong>: <a href=https://github.com/OpenGene/fastp><code>Fastp</code></a> – Trims low-quality reads and adapter sequences.</li> <li><strong>Taxonomic Classification</strong>: <a href=https://ccb.jhu.edu/software/kraken2/ ><code>Kraken2</code></a> – Identifies bacterial species present in the sequencing data.</li> <li><strong>Genome Assembly</strong>: <a href=https://cab.spbu.ru/software/spades/ ><code>SPAdes</code></a> – Performs de novo genome assembly.</li> <li><strong>Quality Assessment of Assembly</strong>: <a href=http://quast.sourceforge.net/ ><code>QUAST</code></a> – Evaluates assembly quality.</li> <li><strong>Plasmid Identification</strong>: <a href=https://cge.food.dtu.dk/services/PlasmidFinder/ ><code>PlasmidFinder</code></a> – Detects plasmids in the assembled genome.</li> <li><strong>MLST Typing</strong>: <a href=https://github.com/tseemann/mlst><code>MLST</code></a> – Determines <strong>Multilocus Sequence Typing (MLST)</strong> for bacterial isolates.</li> <li><strong>AMR Gene Detection</strong>: <a href=https://www.ncbi.nlm.nih.gov/pathogens/antimicrobial-resistance/AMRFinder/ ><code>AMRFinderPlus</code></a> – Identifies antimicrobial resistance genes.</li> <li><strong>Virulence Gene Detection</strong>: <a href=https://github.com/ ><code>Gamma</code></a> – Predicts bacterial virulence factors.</li> <li><strong>Genome Annotation</strong>: <a href=https://github.com/tseemann/prokka><code>Prokka</code></a> – Annotates genomic features.</li> <li><strong>Whole-Genome Comparisons</strong>:<ul> <li><a href=https://github.com/marbl/Mash><code>Mash</code></a> – Estimates genome distances for clustering.</li> <li><a href=https://github.com/ParBLiSS/FastANI><code>FastANI</code></a> – Calculates Average Nucleotide Identity (ANI) for species identification.</li> </ul> </li> <li><strong>Final Report</strong>: <a href=http://multiqc.info/ ><code>MultiQC</code></a> – Summarizes results across all samples in a single interactive report.</li> </ol> <hr> <h2 id=entry-points>Entry Points<a class=headerlink href=#entry-points title="Permanent link">&para;</a></h2> <p>Currently, there are several entry points for the AR pipeline:</p> <ol> <li><code>arBASESPACE</code>: Downloads files directly from Illumina Basespace.</li> <li><code>arANALYSIS</code>: Downloads file Illumina Basespace, if needed; performs <strong>quality control, genome assembly, and taxonomic classification</strong>; detects <strong>antimicrobial resistance genes, plasmids, and virulence factors</strong>; performs <strong>generation of unique WGS ID's</strong>; preparaes <strong>NCBI submission</strong>.</li> <li><code>DBProcessing</code>: Performs <strong>quality control, compiles NCBI ID's into a compiled user file</strong>.</li> </ol> <p>In addition there are several entry points for AR outbreak analysis. 1. <code>outbreakANALYSIS</code>: Performs additional analysis for outbreak detection. 2. <code>outbreakREPORTING</code>: Generates reports depending on the type of outbreak analysis required. 3. <code>NFCORE_OUTBREAK</code>: Executes <strong>outbreakANALYSIS</strong> and <strong>outbreakREPORTING</strong> for an second <strong>end-to-end workflow</strong>.</p> <hr> <h2 id=processes>Processes<a class=headerlink href=#processes title="Permanent link">&para;</a></h2> <h3 id=quality-control-preprocessing><strong>Quality Control &amp; Preprocessing</strong><a class=headerlink href=#quality-control-preprocessing title="Permanent link">&para;</a></h3> <ul> <li><strong>FastQC</strong>: Provides sequencing quality metrics.</li> <li><strong>Fastp</strong>: Trims low-quality reads and removes adapters.</li> <li><strong>Kraken2</strong>: Classifies bacterial species from sequencing data.</li> </ul> <h3 id=genome-assembly-assessment><strong>Genome Assembly &amp; Assessment</strong><a class=headerlink href=#genome-assembly-assessment title="Permanent link">&para;</a></h3> <ul> <li><strong>SPAdes</strong>: Performs de novo genome assembly.</li> <li><strong>QUAST</strong>: Evaluates assembly quality.</li> <li><strong>PlasmidFinder</strong>: Identifies plasmid sequences in assembled genomes.</li> </ul> <h3 id=genotyping-comparative-genomics><strong>Genotyping &amp; Comparative Genomics</strong><a class=headerlink href=#genotyping-comparative-genomics title="Permanent link">&para;</a></h3> <ul> <li><strong>MLST</strong>: Identifies bacterial strain types.</li> <li><strong>Mash</strong>: Estimates genetic distances between isolates.</li> <li><strong>FastANI</strong>: Computes <strong>Average Nucleotide Identity (ANI)</strong> for species classification.</li> </ul> <h3 id=antimicrobial-resistance-virulence-detection><strong>Antimicrobial Resistance &amp; Virulence Detection</strong><a class=headerlink href=#antimicrobial-resistance-virulence-detection title="Permanent link">&para;</a></h3> <ul> <li><strong>AMRFinderPlus</strong>: Detects AMR genes in bacterial genomes.</li> <li><strong>Gamma</strong>: Predicts virulence factors from genomic data.</li> </ul> <h3 id=genome-annotation><strong>Genome Annotation</strong><a class=headerlink href=#genome-annotation title="Permanent link">&para;</a></h3> <ul> <li><strong>Prokka</strong>: Annotates bacterial genomes.</li> </ul> <h3 id=final-reports><strong>Final Reports</strong><a class=headerlink href=#final-reports title="Permanent link">&para;</a></h3> <ul> <li><strong>MultiQC</strong>: Compiles a summary of all quality control and analysis results.</li> </ul> <hr> <h2 id=dependencies><a href=#dependencies>Dependencies</a><a class=headerlink href=#dependencies title="Permanent link">&para;</a></h2> <h3 id=software-tools><strong>Software &amp; Tools</strong><a class=headerlink href=#software-tools title="Permanent link">&para;</a></h3> <p>Ensure that the following software is installed before running the pipeline:</p> <ul> <li><a href=https://www.nextflow.io/docs/latest/getstarted.html#installation><code>Nextflow</code></a> (<code>&gt;=21.04.0</code>)</li> <li><a href=https://docs.docker.com/engine/installation/ ><code>Docker</code></a> or <a href=https://www.sylabs.io/guides/3.0/user-guide/ ><code>Singularity</code></a></li> <li><a href=https://conda.io/miniconda.html><code>Conda</code></a> (Optional)</li> </ul> <h3 id=core-software-versions><strong>Core Software Versions</strong><a class=headerlink href=#core-software-versions title="Permanent link">&para;</a></h3> <p>The pipeline requires the following tools:</p> <div class=highlight><pre><span></span><code>- Python 3.9
- Samtools 1.21
- FastQC 0.12.1
- Fastp 0.23.4
- Kraken2 2.1.2
- SPAdes 3.15.5
- QUAST 5.0.2
- PlasmidFinder 2.1
- MLST 2.23.0
- Mash 2.3
- FastANI 1.33
- AMRFinderPlus 3.10
- Gamma 2.2
- Prokka 1.14.5
- MultiQC 1.21
</code></pre></div> <h3 id=reference-databases>Reference Databases<a class=headerlink href=#reference-databases title="Permanent link">&para;</a></h3> <h4 id=kraken2-db>Kraken2 DB<a class=headerlink href=#kraken2-db title="Permanent link">&para;</a></h4> <p>The database file can be taken from <a href=https://benlangmead.github.io/aws-indexes/k2><code>Ben Langmead's repository</code></a> which links directly to the database file. It is recommended to use the latest version of the <a href=https://genome-idx.s3.amazonaws.com/kraken/k2_standard_08gb_20240904.tar.gz><code>8GB database</code></a>, and reformat it using the <code>bin/reformat_kraken.sh</code> script.</p> <h2 id=pipeline-execution><strong>Pipeline Execution</strong><a class=headerlink href=#pipeline-execution title="Permanent link">&para;</a></h2> <p>The <strong>ODHL_rAcecaR pipeline</strong> is implemented using <strong>Nextflow</strong>, which allows for execution on <strong>local machines, HPC clusters, or cloud environments</strong>.</p> <h3 id=1-install-nextflow><strong>1. Install Nextflow</strong><a class=headerlink href=#1-install-nextflow title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code>curl<span class=w> </span>-s<span class=w> </span>https://get.nextflow.io<span class=w> </span><span class=p>|</span><span class=w> </span>bash
mv<span class=w> </span>nextflow<span class=w> </span>~/bin/
</code></pre></div> <h3 id=2-clone-the-repository><strong>2. Clone the Repository</strong><a class=headerlink href=#2-clone-the-repository title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code>git<span class=w> </span>clone<span class=w> </span>https://github.com/ODHL/ODHL_rAcecaR.git
<span class=nb>cd</span><span class=w> </span>ODHL_rAcecaR
</code></pre></div> <h3 id=3-configure-the-pipeline><strong>3. Configure the Pipeline</strong><a class=headerlink href=#3-configure-the-pipeline title="Permanent link">&para;</a></h3> <p>Modify the <strong>Nextflow configuration file (<code>nextflow.config</code>)</strong> to specify reference databases and execution profiles.</p> <p>Example modification: <div class=highlight><pre><span></span><code><span class=n>params</span><span class=w> </span><span class=o>{</span>
<span class=w>    </span><span class=n>kraken2_db</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s2>&quot;/home/ubuntu/refs/k2/k2_standard_08gb_202412.tar.gz&quot;</span>
<span class=w>    </span><span class=n>amrfinder_db</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s2>&quot;/home/ubuntu/refs/amrfinderplus/latest&quot;</span>
<span class=w>    </span><span class=n>plasmidfinder_db</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s2>&quot;/home/ubuntu/refs/plasmidfinder/latest&quot;</span>
<span class=o>}</span>
</code></pre></div></p> <h3 id=4-optional-install-basespace>*** 4. (Optional) Install Basespace**<a class=headerlink href=#4-optional-install-basespace title="Permanent link">&para;</a></h3> <p>The pipeline allows for automatic download from basespace. If you choose to use this feature, you'll need to add basespace to your $PATH. <div class=highlight><pre><span></span><code><span class=c1># Install basespace</span>
<span class=c1>## Docs</span>
<span class=c1>## https://developer.basespace.illumina.com/docs/content/documentation/cli/cli-overview</span>
<span class=k>if</span><span class=w> </span><span class=o>[[</span><span class=w> </span>!<span class=w> </span>-d<span class=w> </span><span class=nv>$HOME</span>/tools/<span class=w> </span><span class=o>]]</span><span class=p>;</span><span class=w> </span><span class=k>then</span><span class=w> </span>mkdir<span class=w> </span>-p<span class=w> </span><span class=nv>$HOME</span>/tools/<span class=p>;</span><span class=w> </span><span class=k>done</span>
wget<span class=w> </span><span class=s2>&quot;https://launch.basespace.illumina.com/CLI/latest/amd64-linux/bs&quot;</span><span class=w> </span>-O<span class=w> </span><span class=nv>$HOME</span>/tools/basespace
chmod<span class=w> </span>u+x<span class=w> </span><span class=nv>$HOME</span>/tools/basespace
./basespace<span class=w> </span>auth
<span class=c1>### follow path to website and sign in</span>
<span class=c1>### should display &quot;Welcome [name of user]</span>
</code></pre></div></p> <hr> <h2 id=reproducibility><strong>Reproducibility</strong><a class=headerlink href=#reproducibility title="Permanent link">&para;</a></h2> <p>To ensure <strong>consistent results</strong>, specify the pipeline version when running:</p> <div class=highlight><pre><span></span><code>nextflow<span class=w> </span>run<span class=w> </span>ODHL/ODHL_AR<span class=w> </span>-r<span class=w> </span><span class=m>1</span>.0.0
</code></pre></div> <p>You can check for the latest version on the <strong><a href=https://github.com/ODHL/ODHL_AR/releases/ >ODHL/ODHL_AR GitHub Releases page</a></strong>.</p> <hr> <hr> <div class=md-source-file> <small> Last update: 2025-01-29 </small> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2025 ODHL </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href="https://odh.ohio.gov/about-us/offices-bureaus-and-departments/bphl/welcome-to?contentIDR=9ce211e2-7949-447b-bf0e-2abfd5da7892&useDefaultText=1&useDefaultDesc=0" target=_blank rel=noopener title=odh.ohio.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M144 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zm368 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zM0 298.7C0 239.8 47.8 192 106.7 192h42.7c15.9 0 31 3.5 44.6 9.7-1.3 7.2-1.9 14.7-1.9 22.3 0 38.2 16.8 72.5 43.3 96H21.3C9.6 320 0 310.4 0 298.7zM405.3 320h-.7c26.6-23.5 43.3-57.8 43.3-96 0-7.6-.7-15-1.9-22.3 13.6-6.3 28.7-9.7 44.6-9.7h42.7c58.9 0 106.7 47.8 106.7 106.7 0 11.8-9.6 21.3-21.3 21.3H405.3zM224 224a96 96 0 1 1 192 0 96 96 0 1 1-192 0zm-96 261.3c0-73.6 59.7-133.3 133.3-133.3h117.4c73.6 0 133.3 59.7 133.3 133.3 0 14.7-11.9 26.7-26.7 26.7H154.7c-14.7 0-26.7-11.9-26.7-26.7z"/></svg> </a> <a href=https://github.com/ODHL target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.tabs", "navigation.top", "navigation.indexes", "toc.integrate"], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.51198bba.min.js></script> </body> </html>