# **rAcecaR: Antimicrobial Resistance (AR) Bioinformatics Pipeline**

[![Hits](https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fgithub.com%2FODHL%2FODHL_AR%2F&count_bg=%233D4FC8&title_bg=%23555555&icon=&icon_color=%23E7E7E7&title=PageHits&edge_flat=false)](https://hits.seeyoufarm.com)

ðŸš¨ **Warning** ðŸš¨  
The results generated by this pipeline are **not CLIA certified** and should not be considered diagnostic.

---

## **Introduction**  
The **ODHL_rAcecaR pipeline** is designed for the detection and characterization of antimicrobial resistance (AMR) genes and other relevant genomic features from high-throughput sequencing data. This pipeline processes bacterial whole genome sequencing (WGS) and metagenomic sequencing data to identify antimicrobial resistance determinants, sequence types, and genomic features relevant to public health surveillance and epidemiological investigations.

This workflow is developed and maintained by the [`Ohio Department of Health Laboratory (ODHL)`](https://odh.ohio.gov/about-us/offices-bureaus-and-departments/bphl/welcome-to?contentIDR=9ce211e2-7949-447b-bf0e-2abfd5da7892&useDefaultText=1&useDefaultDesc=0) and is built using [`Nextflow`](https://www.nextflow.io/), a robust workflow management system that ensures reproducibility and scalability across HPC clusters, cloud environments, and local machines. The pipeline was built off the work of the [`CDC Phoenix Pipeline`](https://github.com/CDCgov/phoenix). The workflow leverages [`Docker`](https://www.docker.com/) and [`Singularity containers`](https://www.singularity.com/) to provide a standardized execution environment.
[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A521.10.3-23aa62.svg?labelColor=000000)](https://www.nextflow.io/)

---

## **Pipeline Workflow**
The **ODHL_rAcecaR pipeline** consists of the following stages:

### **1. Read Quality Control (QC)**
- **Tool:** [`FastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)  
- **Purpose:** Evaluates sequencing read quality for potential issues such as adapter contamination and sequencing errors.

### **2. Read Trimming and Filtering**
- **Tool:** [`Fastp`](https://github.com/OpenGene/fastp)  
- **Purpose:** Removes low-quality bases and adapter sequences to improve downstream analysis.

### **3. Taxonomic Classification**
- **Tool:** [`Kraken2`](https://ccb.jhu.edu/software/kraken2/)  
- **Purpose:** Assigns taxonomic labels to sequencing reads based on a comprehensive reference database.

- **Tool:** [`Krona`](https://github.com/marbl/Krona)  
- **Purpose:** Generates interactive visualizations of taxonomic classification results.

### **4. Genome Assembly and Quality Assessment**
- **Tool:** [`SPAdes`](https://github.com/ablab/spades)  
- **Purpose:** Performs **de novo genome assembly** from short-read sequencing data.

- **Tool:** [`QUAST`](https://github.com/ablab/quast)  
- **Purpose:** Assesses the quality and completeness of genome assemblies.

### **5. Sequence Typing & Resistance Gene Detection**
- **Tool:** [`MLST`](https://github.com/tseemann/mlst)  
- **Purpose:** Identifies **sequence types (STs)** of bacterial genomes for epidemiological tracking.

- **Tool:** [`AMRFinderPlus`](https://www.ncbi.nlm.nih.gov/pathogens/antimicrobial-resistance/AMRFinder/)  
- **Purpose:** Detects known **antimicrobial resistance (AMR) genes and point mutations** associated with resistance.

### **6. Plasmid & Virulence Gene Identification**
- **Tool:** [`PlasmidFinder`](https://bitbucket.org/genomicepidemiology/plasmidfinder/src/master/)  
- **Purpose:** Identifies plasmid sequences associated with resistance gene transmission.

- **Tool:** [`Gamma`](https://github.com/CFSAN-Biostatistics/gamma)  
- **Purpose:** Detects horizontally transferred genes associated with **virulence and resistance**.

### **7. Phylogenetic Analysis & Similarity Metrics**
- **Tool:** [`FastANI`](https://github.com/ParBLiSS/FastANI)  
- **Purpose:** Measures genome-wide similarity between bacterial isolates.

- **Tool:** [`Mash`](https://github.com/marbl/Mash)  
- **Purpose:** Quickly estimates genetic distance between samples for clustering analyses.

---

## **Input Requirements**
The **ODHL_rAcecaR** pipeline processes **paired-end or single-end Illumina sequencing reads** in **FASTQ format**. A **sample sheet (CSV/TSV)** should be provided with metadata for each sample with other input. View the  [`Getting Started page`](https://odhl.github.io/odhl_ar/user-guide/getting-started/) and [`Preparing Files page`](https://odhl.github.io/odhl_ar/user-guide/preparing-files/) for more information.for more information.

---

## **Output Summary**
View the  [`Output page`](https://odhl.github.io/odhl_ar/user-guide/getting-started/) for more information.

---

## **Citations & Acknowledgments**
This pipeline was developed by the **Ohio Department of Health Laboratory (ODHL)** to support **public health surveillance of antimicrobial resistance (AMR)**. For contribution and contact information, view the [`contribution page`](https://odhl.github.io/odhl_ar/user-guide/contributions/).